spec/middleware-1.0.0-draft.1.md

```diff
@@ -0,0 +1,313 @@
+
+ - **Builder**: The extension point lambda used by developers to add Middleware
+   components to a pipeline.
+
+ - **Server**: xxx
+
+
+## 3. Middleware
+
+An OWIN Application (as defined by an AppFunc) can be modeled as
+Middleware components layered in a pipeline.
+
+Each Middleware receiving a request may perform some processing on the request,
+may forward the request to the next Middleware and return the resulting
+response, or may decide to return a response immediately and not call the rest
+of the chain.
+
+In other words, a Middleware functions as a proxy of sort, completely
+encapsulating and hiding the next Middleware in the chain, transparently to the
+caller.
+
+### 3.1. Examples
+
+For example, a Middleware responsible for logging all requests and responses
+could execute as the first component of a pipeline, and log all requests and
+responses as they pass through.
+
+```
+        request      > (log request)      > (process request)
+SERVER =============== LOGGER ============= WEB APPLICATION
+        response     < (log response)     < (return response)
+```
+
+Other types of Middleware would respond immediately, without calling the next
+component in the pipeline, such as for an authorization module denying a request
+before a web application can respond.
+
+```
+        request      > (authenticate)
+SERVER =============== LOGGER ------X------ WEB APPLICATION
+        response     < (deny)
+```
+
+### 3.2. Middleware Operation
+
+The whole set of Middleware composing an Application are first chained together
+in a pipeline in a composition phase. The result of Middleware composition is an
+Application (OWIN AppFunc).
+
+A Server as defined by OWIN will call the resulting Application, unaware of the
+composition of the pipeline it is calling.
+
+At compose time, each Middleware is passed a reference to the next Middleware in
+the chain, and should return an Application.
+
+At run time, a Middleware can inspect and manipulate HTTP messages using the
+usual environment dictionary, and in by doing so leverages existing knowledge in
+building OWIN-compliant components.
+
+### 3.3. Signature
+
+```csharp
+using MidFunc = Func<
+                     AppFunc, // next Middleware in the chain
+                     AppFunc // returns the resulting application
+                    >
+```
+
+### 3.4. Code Sample
+
+The following example is a Middleware logging the request and response headers.
+
+```csharp
+// MidFunc
+public static AppFunc LogMiddleware(AppFunc nextMiddleware)
+{
+  AppFunc composedApplication = environment => {
+
+    // log the request headers
+    Log.RequestHeaders(environment["owin.RequestHeaders"]);
+
+    // pass the request to the next Middleware
+    next(environment);
+
+    // log the response headers
+    Log.RequestHeaders(environment["owin.ResponseHeaders"]);
+
+  }
+
+  return composedApplication;
+}
+```
+
+> **Note**
+>
+This specification does not preclude alternative modeling of the Middleware
```

Comment:
I think it would be nicer if blockquotes started with `>` on all lines. See http://tirania.org/blog/archive/2014/Sep.html